myst_nbs/ode_models/ODE_Lotka_Volterra_multiple_ways.myst.md

```diff
@@ -0,0 +1,964 @@
+# ODE Lotka-Volterra With Bayesian Inference in Multiple Ways
+
+:::{post} December 27, 2022
+:tags: ODE, aesara op, gradient-free inference, aesara scan
+:category: intermediate
+:type: how-to
+:author: Adapted by Greg Brunkhorst from multiple pymc3 example notebooks by Sanmitra Ghosh, Demetri Pananos, and the PyMC Team
+:::
+
+```{code-cell} ipython3
+import aesara
+import aesara.tensor as at
+import arviz as az
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pymc as pm
+
+from aesara.compile.ops import as_op
+from pymc.ode import DifferentialEquation
+from scipy.integrate import odeint
+from scipy.optimize import least_squares
+
+print(f"Running on PyMC v{pm.__version__}")
+```
+
+```{code-cell} ipython3
+%load_ext watermark
+az.style.use("arviz-darkgrid")
+rng = np.random.default_rng(1234)
+```
+
+## Purpose
+The purpose of this notebook is to demonstrate how to perform Bayesian inference on a system of ordinary differential equations (ODEs), both with and without gradients.  The accuracy and efficiency of different samplers are compared.    
+### Outline
+* Background
+    * Motivation
+    * Lotka-Volterra Predator Prey Model
+    * Hudson's Bay Company Data
+* Scipy `odeint`
+    * Specification
+    * Least Squares Solution
+* Gradient-free Bayesian Inference
+    * Wrap `odeint` in an Aesara operator for use in PyMC
+    * Bayesian inference using gradient-free methods
+        * Slice Sampler
+        * DEMetropolisZ Sampler
+        * DEMetropolis Sampler
+        * Metropolis Sampler
+        * Sequential Monte Carlo (SMC) Sampler
+* Bayesian Inference with Gradients
+    * `pymc.ode.DifferentialEquation` specification with the NUTs Sampler
+    * Looping in PyMC with `aesara.scan` and the NUTs Sampler
+### Key Conclusions
+* Based on the experiments in this notebook, the most simple and efficient method for performing Bayesian inference on the Lotka-Volterra equations was to specify the ODE system in Scipy, wrap the function as an Aesara op, and use a Differential Evolution Metropolis (DEMetropolis) sampler in PyMC.  
+
++++ {"tags": []}
+
+## Background
+### Motivation
+Ordinary differential equation models (ODEs) are used in a variety of science and engineering domains to model the time evolution of physical variables.  A natural choice to estimate the values and uncertainty of model parameters given experimental data is Bayesian inference.  However, ODEs can be challenging to specify and solve in the Bayesian setting, therefore, this notebook steps through multiple methods for solving an ODE inference problem using PyMC. The Lotka-Volterra model used in this example has often been used for benchmarking Bayesian inference methods (e.g., in [Stan](https://mc-stan.org/users/documentation/case-studies/lotka-volterra-predator-prey.html)).
+
+### Lotka-Volterra Predator-Prey Model
+The Lotka-Volterra model describes the interaction between a predator and prey species. This ODE given by:  
+$$
```

Comment:
```suggestion
The Lotka-Volterra model describes the interaction between a predator and prey species. This ODE given by:

$$
```

Commenting here as it is much easier to review these issues, but you should make the changes on the jupyter notebook, don't edit the myst file directly.

I think you'll need to add this empty line for rendering to work properly on the website. It still looks like inline math: https://pymcio--478.org.readthedocs.build/projects/examples/en/478/ode_models/ODE_Lotka_Volterra_multiple_ways.html#lotka-volterra-predator-prey-model